packages/react-list/Spec.md

```diff
@@ -2,62 +2,375 @@
+
+  return <BasicList componentRef={ref}>;
+}
+```
 
 ## Sample Code
 
-_Provide some representative example code that uses the proposed API for the component_
+Presentational/Non-interactive list
+
+```tsx
+<List>
+  <ListItem>List Item</ListItem>
+  <ListItem>List Item</ListItem>
+  <ListItem>List Item</ListItem>
+  <ListItem>List Item</ListItem>
+</List>
+```
+
+Selectable listbox
+
+```tsx
+<Listbox>
+  <ListboxOption value="1">Option 1</ListboxOption>
+  <ListboxOption value="2">Option 2</ListboxOption>
+  <ListboxOption value="3">Option 3</ListboxOption>
+  <ListboxOption value="4">Option 4</ListboxOption>
+</Listbox>
+```
+
+Selectable listbox without checkbox indicator slot
+
+```tsx
+<Listbox>
+  <ListboxOption checkbox={null} value="2">
+    Option 1
+  </ListboxOption>
+  <ListboxOption checkbox={null} value="2">
+    Option 2
+  </ListboxOption>
+  <ListboxOption checkbox={null} value="3">
+    Option 3
+  </ListboxOption>
+  <ListboxOption checkbox={null} value="4">
+    Option 4
+  </ListboxOption>
+</Listbox>
+```
+
+Selectable listbox with multiselect
+
+```tsx
+<Listbox multiselect>
+  <ListboxOption value="1">Option 1</ListboxOption>
+  <ListboxOption value="2">Option 2</ListboxOption>
+  <ListboxOption value="3">Option 3</ListboxOption>
+  <ListboxOption value="4">Option 4</ListboxOption>
+</Listbox>
+```
 
 ## Variants
 
-_Describe visual or functional variants of this control, if applicable. For example, a slider could have a 2D variant._
+### Single select
+
+The selectable list should only select one single element at a time when invoked. This functions similar to a
+radio group behavior.
+
+### Multi select
+
+The selectable list should toggle the selection state of each item when invoked. This is similar to the behaviour of a
+group of checkboxes.
 
 ## API
 
-_List the **Props** and **Slots** proposed for the component. Ideally this would just be a link to the component's `.types.ts` file_
+### List
+
+```ts
+type ListSlots = {
+  root: NonNullable<Slot<'ul' | 'div'>>;
+};
+
+type ListProps = ComponentProps<ListSlots>;
+```
+
+### ListItem
+
+```ts
+type ListItemSlots = {
+  root: NonNullable<Slot<'li' | 'div'>>;
+};
+
+type ListItemProps = ComponentProps<ListItemSlots>;
+```
+
+### Listbox
+
+```ts
+type ListboxSlots = ListSLots;
+
+type ListboxCommons = {
+  onChange: (e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>, data: { selectedValues: string[] });
+  selectedValues: string[];
+  defaultSelectedValues: string[];
+}
+
+type ListboxProps = ComponentProps<ListboxSlots> & ListboxCommons;
+```
+
+### ListboxOption
+
+```ts
+type ListboxOptionSlots = ListItemSlots & {
+  checkbox: Slot<'span'>;
+};
+
+type ListboxOptionCommons = {
+  value: string;
+};
+
+type ListboxOptionProps = ComponentProps<ListboxOptionSlots> & ListboxOptionCommons;
+```
 
 ## Structure
 
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+### List
+
+```tsx
+<List>
```

Comment:
I think this question is intertwined with virtualization but do you think we'll want to support an `items` prop similar to v0/8 in addition to children like you've shown here?

A list with hundreds or thousands of items seems reasonable and requiring users to iterate them to produce children seems like extra work and could end up being quite complicated when combined with a virtualization solution.

Reply:
I'm not sure how using `children` becomes a problem here. Virtualization libraries normally provide a simple way of adapting to this pattern.

Take `react-window` for example:

```tsx
import { FixedSizeList } from "react-window";

const Row = ({ index, style }) => (
  <ListItem style={style}>item {index}</ListItem>
);

const Example = () => (
  <FixedSizeList
    height={150}
    // by using innerElementType we can override the base element for the FluentUI List
    innerElementType={innerElementType}
    itemCount={51}
    // We'll need to somehow provide the size of an ListItem, so that this is possible
    itemSize={ITEM_SIZE}
    width={300}
  >
    {Row}
  </FixedSizeList>
);

const innerElementType = forwardRef((props, ref) => (
  <List ref={ref} {...props} />
));

```


Although some libraries. might not present such functionality, for example [resembli](https://www.react-virtualized-window.com/list#props), in that case I'd suggest we document a way of how we can use our compositional API to actually remount `List` and `ListItem` components on a proper way that would be usable for such libraries:

```tsx
import { List as VirtualizedList } from "@resembli/react-virtualized-window";
import {
  ListItem,
  LIST_ITEM_SIZE,
  useList,
  useListStyle,
  renderList
} from "@fluentui/react-list";

const Row = ({ data, style }) => <ListItem style={style}>{data}</ListItem>;

export const List = React.forwardRef((props, ref) => {
  const state = useList(props, ref);
  state.components.root = VirtualizedList;
  state.root = { ...state.root, size: LIST_ITEM_SIZE, data: props.data };
  useListStyle(state);
  return renderList(props);
});

export function Example() {
  return (
    <div style={{ width: "100%", height: 400 }}>
      <List data={someData}>{Row}</List>
    </div>
  );
}
```